File: go/compress.go
package ics23

import "fmt"

// IsCompressed returns true if the proof was compressed
func IsCompressed(proof *CommitmentProof) bool {
	return proof.GetCompressed() != nil
}

// Compress will return a CompressedBatchProof if the input is BatchProof
// Otherwise it will return the input.
// This is safe to call multiple times (idempotent)
func Compress(proof *CommitmentProof) *CommitmentProof {
	batch := proof.GetBatch()
	if batch == nil {
		return proof
	}
	return &CommitmentProof{
		Proof: &CommitmentProof_Compressed{
			Compressed: compress(batch),
		},
	}
}

// Decompress will return a BatchProof if the input is CompressedBatchProof
// Otherwise it will return the input.
// This is safe to call multiple times (idempotent)
func Decompress(proof *CommitmentProof) (*CommitmentProof, error) {
	comp := proof.GetCompressed()
	if comp != nil {
		batch, err := decompress(comp)
		if err != nil {
			return nil, err
		}
		return &CommitmentProof{
			Proof: &CommitmentProof_Batch{
				Batch: batch,
			},
		}, nil
	}
	return proof, nil
}

func compress(batch *BatchProof) *CompressedBatchProof {
	var centries []*CompressedBatchEntry
	var lookup []*InnerOp
	registry := make(map[string]int32)

	for _, entry := range batch.Entries {
		centry := compressEntry(entry, &lookup, registry)
		centries = append(centries, centry)
	}

	return &CompressedBatchProof{
		Entries:      centries,
		LookupInners: lookup,
	}
}

func compressEntry(entry *BatchEntry, lookup *[]*InnerOp, registry map[string]int32) *CompressedBatchEntry {
	if exist := entry.GetExist(); exist != nil {
		return &CompressedBatchEntry{
			Proof: &CompressedBatchEntry_Exist{
				Exist: compressExist(exist, lookup, registry),
			},
		}
	}

	non := entry.GetNonexist()
	return &CompressedBatchEntry{
		Proof: &CompressedBatchEntry_Nonexist{
			Nonexist: &CompressedNonExistenceProof{
				Key:   non.Key,
				Left:  compressExist(non.Left, lookup, registry),
				Right: compressExist(non.Right, lookup, registry),
			},
		},
	}
}

func compressExist(exist *ExistenceProof, lookup *[]*InnerOp, registry map[string]int32) *CompressedExistenceProof {
	if exist == nil {
		return nil
	}
	res := &CompressedExistenceProof{
		Key:   exist.Key,
		Value: exist.Value,
		Leaf:  exist.Leaf,
		Path:  make([]int32, len(exist.Path)),
	}
	for i, step := range exist.Path {
		res.Path[i] = compressStep(step, lookup, registry)
	}
	return res
}

func compressStep(step *InnerOp, lookup *[]*InnerOp, registry map[string]int32) int32 {
	bz, err := step.Marshal()
	if err != nil {
		panic(err)
	}

	// load from cache if there
	if num, ok := registry[string(bz)]; ok {
		return num
	}

	// create new step if not there
	num := int32(len(*lookup))
	*lookup = append(*lookup, step)
	registry[string(bz)] = num
	return num
}

func decompress(comp *CompressedBatchProof) (*BatchProof, error) {
	lookup := comp.LookupInners

	var entries []*BatchEntry

	for _, centry := range comp.Entries {
		entry, err := decompressEntry(centry, lookup)
		if err != nil {
			return nil, err
		}
		entries = append(entries, entry)
	}

	return &BatchProof{
		Entries: entries,
	}, nil
}

func decompressEntry(entry *CompressedBatchEntry, lookup []*InnerOp) (*BatchEntry, error) {
	if exist := entry.GetExist(); exist != nil {
		decompressedExist, err := decompressExist(exist, lookup)
		if err != nil {
			return nil, err
		}
		return &BatchEntry{
			Proof: &BatchEntry_Exist{
				Exist: decompressedExist,
			},
		}, nil
	}

	non := entry.GetNonexist()
	decompressedLeft, err := decompressExist(non.Left, lookup)
	if err != nil {
		return nil, err
	}

	decompressedRight, err := decompressExist(non.Right, lookup)
	if err != nil {
		return nil, err
	}

	return &BatchEntry{
		Proof: &BatchEntry_Nonexist{
			Nonexist: &NonExistenceProof{
				Key:   non.Key,
				Left:  decompressedLeft,
				Right: decompressedRight,
			},
		},
	}, nil
}

func decompressExist(exist *CompressedExistenceProof, lookup []*InnerOp) (*ExistenceProof, error) {
	if exist == nil {
		return nil, nil
	}
	res := &ExistenceProof{
		Key:   exist.Key,
		Value: exist.Value,
		Leaf:  exist.Leaf,
		Path:  make([]*InnerOp, len(exist.Path)),
	}
	for i, step := range exist.Path {
		if int(step) >= len(lookup) {
			return nil, fmt.Errorf("compressed existence proof at index %d has lookup index out of bounds", i)
		}
		res.Path[i] = lookup[step]
	}
	return res, nil
}

File: js/src/compress.ts
import { ics23 } from "./generated/codecimpl";

export function compress(
  proof: ics23.ICommitmentProof
): ics23.ICommitmentProof {
  if (!proof.batch) {
    return proof;
  }
  return { compressed: compressBatch(proof.batch) };
}

export function decompress(
  proof: ics23.ICommitmentProof
): ics23.ICommitmentProof {
  if (!proof.compressed) {
    return proof;
  }
  return { batch: decompressBatch(proof.compressed) };
}

function compressBatch(proof: ics23.IBatchProof): ics23.ICompressedBatchProof {
  const centries: ics23.ICompressedBatchEntry[] = [];
  const lookup: ics23.IInnerOp[] = [];
  const registry = new Map<Uint8Array, number>();

  for (const entry of proof.entries!) {
    if (entry.exist) {
      const centry = { exist: compressExist(entry.exist, lookup, registry) };
      centries.push(centry);
    } else if (entry.nonexist) {
      const non = entry.nonexist;
      const centry = {
        nonexist: {
          key: non.key,
          left: compressExist(non.left, lookup, registry),
          right: compressExist(non.right, lookup, registry),
        },
      };
      centries.push(centry);
    } else {
      throw new Error("Unexpected batch entry during compress");
    }
  }

  return {
    entries: centries,
    lookupInners: lookup,
  };
}

function compressExist(
  exist: ics23.IExistenceProof | null | undefined,
  lookup: ics23.IInnerOp[],
  registry: Map<Uint8Array, number>
): ics23.ICompressedExistenceProof | undefined {
  if (!exist) {
    return undefined;
  }

  const path = exist.path!.map((inner) => {
    const sig = ics23.InnerOp.encode(inner).finish();
    let idx = registry.get(sig);
    if (idx === undefined) {
      idx = lookup.length;
      lookup.push(inner);
      registry.set(sig, idx);
    }
    return idx;
  });

  return {
    key: exist.key,
    value: exist.value,
    leaf: exist.leaf,
    path,
  };
}

function decompressBatch(
  proof: ics23.ICompressedBatchProof
): ics23.IBatchProof {
  const lookup = proof.lookupInners!;
  const entries = proof.entries!.map((comp) => {
    if (comp.exist) {
      return { exist: decompressExist(comp.exist, lookup) };
    } else if (comp.nonexist) {
      const non = comp.nonexist;
      return {
        nonexist: {
          key: non.key,
          left: decompressExist(non.left, lookup),
          right: decompressExist(non.right, lookup),
        },
      };
    } else {
      throw new Error("Unexpected batch entry during compress");
    }
  });
  return {
    entries,
  };
}

function decompressExist(
  exist: ics23.ICompressedExistenceProof | null | undefined,
  lookup: readonly ics23.IInnerOp[]
): ics23.IExistenceProof | undefined {
  if (!exist) {
    return undefined;
  }
  const { key, value, leaf, path } = exist;
  const newPath = (path || []).map((idx) => lookup[idx]);
  return { key, value, leaf, path: newPath };
}
